build-logic/src/main/kotlin/dokkabuild/utils/gradle.kt

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package dokkabuild.utils
+
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.attributes.*
+import org.gradle.api.attributes.java.TargetJvmEnvironment
+import org.gradle.api.model.ObjectFactory
+import org.gradle.kotlin.dsl.named
+
+
+/**
+ * Mark this [Configuration] as one that should be used to declare dependencies in
+ * [org.gradle.api.Project.dependencies] block.
+ *
+ * Declarable Configurations should be extended by [resolvable] and [consumable] Configurations.
+ * They must not have attributes.
+ *
+ * ```
+ * isCanBeResolved = false
+ * isCanBeConsumed = false
+ * isCanBeDeclared = true
+ * ```
+ */
+internal fun Configuration.declarable(
```

Comment:
I believe we can already use newer APIs introduced in Gradle 8.4 for this, no?
It's for our build-logic only, so it's fine.
Also, KGP is already using them even under the hood [here](https://github.com/JetBrains/kotlin/blob/0df3a37303f95e86608aa5249115e11ea754c34f/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/utils/configurations.kt#L79), so this should be fine overall.

Reply:
We could do, but I'd recommend waiting for now. I suspect that these new role-based Configurations will be changed in the future, because they're a bit more rough around the edges.

For example, there's no nice Kotlin delegated DSL for creating them, so you have to duplicate the names.


```kotlin
val fooConfig by configurations.resolvable("fooConfig") { 
}
```

And aside from being mentioned in the release notes, there's no documentation for them. 

They probably work fine, but then, so does the existing set up :)

But if you insist, I can update this PR to use the role-based Configs.

Reply:
Ok, no need for now then.